app/gui2/src/components/GraphEditor/collapsing.ts

```diff
@@ -105,21 +107,107 @@ export function prepareCollapsedInfo(selected: Set<ExprId>, graphDb: GraphDb): C
     refactored: {
       id: output.node,
       pattern,
-      expression: 'Main.collapsed' + (inputs.length > 0 ? ' ' : '') + inputs.join(' '),
+      arguments: inputs,
     },
   }
 }
 
-// === performRefactoring ===
+/** Generate a safe method name for a collapsed function using `baseName` as a prefix. */
+function findSafeMethodName(module: Ast.Module, baseName: string): string {
+  const allIdentifiers = moduleMethodNames(module)
+  if (!allIdentifiers.has(baseName)) {
+    return baseName
+  }
+  let index = 1
+  while (allIdentifiers.has(`${baseName}${index}`)) {
+    index++
+  }
+  return `${baseName}${index}`
+}
+
+// === performCollapse ===
+
+// We support working inside `Main` module of the project at the moment.
+const MODULE_NAME = 'Main'
+const COLLAPSED_FUNCTION_NAME = 'collapsed'
 
 /** Perform the actual AST refactoring for collapsing nodes. */
-export function performCollapse(_info: CollapsedInfo) {
-  // The general flow of this function:
-  // 1. Create a new function with a unique name and a list of arguments from the `ExtractedInfo`.
-  // 2. Move all nodes with `ids` from the `ExtractedInfo` into this new function. Use the order of their original definition.
-  // 3. Use a single identifier `output.identifier` as the return value of the function.
-  // 4. Change the expression of the `RefactoredInfo.id` node to the `RefactoredINfo.expression`
-  throw new Error('Not yet implemented, requires AST editing.')
+export function performCollapse(
+  info: CollapsedInfo,
+  module: Ast.Module,
+  topLevel: Ast.BodyBlock,
+  db: GraphDb,
+  currentMethodName: string,
+): Ast.MutableModule {
+  const functionAst = Ast.findModuleMethod(module, currentMethodName)
+  if (!(functionAst instanceof Ast.Function) || !(functionAst.body instanceof Ast.BodyBlock)) {
+    throw new Error(`Expected a collapsable function, found ${functionAst}.`)
+  }
+  const functionBlock = functionAst.body
+  const posToInsert = findInsertionPos(module, topLevel, currentMethodName)
+  const collapsedName = findSafeMethodName(module, COLLAPSED_FUNCTION_NAME)
+  const astIdsToExtract = new Set(
+    [...info.extracted.ids].map((nodeId) => db.nodeIdToNode.get(nodeId)?.outerExprId),
+  )
+  const astIdToReplace = db.nodeIdToNode.get(info.refactored.id)?.outerExprId
+  const collapsed = []
+  const refactored = []
+  const edit = module.edit()
+  const lines = functionBlock.lines()
+  for (const line of lines) {
+    const astId = line.expression?.node.exprId
+    const ast = astId != null ? module.get(astId) : null
+    if (ast == null) continue
+    if (astIdsToExtract.has(astId)) {
+      collapsed.push(ast)
+      if (astId === astIdToReplace) {
+        const ast = collapsedCallAst(info, collapsedName, edit)
```

Comment:
Can we call this var `newAst`? I see lines 165 and 168 are exactly the same, but logically do different things.

Reply:
Done